Adding an analysis flow lesson (#274)
lhcb/starterkit-lessons: first-analysis-steps/analysisflow.md

This is done using the software package called DaVinci.

{% endcallout %}

### Getting data files

After preselecting data either in the stripping, sprucing or triggering step, user can produce a data file, called ntuple by running DaVinci package.
The ntuple is a data file that contains information about events or candidates in the data sample, like mass of a candidate or trigger decision flags.
For the data sample information is coming from reconstruction, but for the simulation sample you can request both the reconstructed values of observables/variables or generator-level values.
The reconstructed values are the ones that passed the entire reconstruction pipeline with detector effects included and generator-level values are the original values that the event generator has produced.
This allows to study different reconstruction and detector effects for your analysis.

DaVinci can be run using CERN grid computing system directly via ganga or via Analysis Productions. 
Using gange will give you more control over the jobs (little scrips and macros you run on the remote machines) and also will allow you to do way more than just running DaVinci. 

> Analysis Productions, however, are more user-friendly and provide

Author: Hi, @vlisovsk ! Thanks for your comments. About the typical analysis flow it is really hard for me to judge the "typical" here. I can write something along the lines: 1. You get Monte Carlo samples and data samples. 2. You make your selections 3. You calibrate your samples 4. You compute acceptances, efficiencies 5. You compute the final quantity. 6. You write ANA note 7. You publish :D

Analysis Productions are also extremely important for the preservation of your analysis.
We will discuss the concept of analysis preservation a bit later in this lesson.
Unless you want to run some really unusual ntuple jobs on ganga, you should always go for Analysis Productions. 

> In first analysis steps we cover both running DaVinci on [ganga](https://lhcb.github.io/starterkit-lessons/first-analysis-steps/davinci-grid.html) and via [Analysis Productions](https://lhcb.github.io/starterkit-lessons/first-analysis-steps/analysis-productions.html).

Reviewer: okay i just proposed to link them above :) feel free to choose.

### Useful high energy physics analysis tools

After getting the ntuples a user usually develops new analysis code or expands an existing code, that their collaborators use. 
Analysis code is usually based on the popular high-energy physics software tools or on the more general data analysis tools, like [numpy](https://numpy.org/) or [pandas](https://pandas.pydata.org/). 

{% callout "Popular data analysis tools in HEP" %}

> There are multiple packages that collaboratos use in LHCb.

Reviewer: ```suggestion
There are multiple packages that collaborators use in LHCb.
```

> Some of this packages are CERN supported, some are projects from the LHCb collaborators and some are supported by external community.  

Reviewer: to not offend anyone (🤪) it would be good to say something like "this list is by no means exhaustive"

Author: Good point 😅

> * [ROOT](https://root.cern/) and it's python twin sister pyROOT are open-source data analysis frameworks used extensively in HEP community. 

Reviewer: ```suggestion
* [ROOT](https://root.cern/) and its python twin sister pyROOT are open-source data analysis frameworks used extensively in HEP community. 
```

> * [scikit-HEP project](https://github.com/scikit-hep) a metaproject for python based particle physics analysis. Consists of multiple useful python modules, some of them are listed here:

Reviewer 2: ```suggestion
* [Scikit-HEP project](https://github.com/scikit-hep), a community project for a Python-based Big Data analysis ecosystem for Particle Physics. Consists of multiple complementary and interoperable Python modules, some of them are listed here:
```

   - [uproot](https://github.com/scikit-hep/uproot4) a python module that allows to read the root files (that are the output of the DaVinci package). 

>    - [awkward arrays](https://github.com/scikit-hep/awkward-1.0) a python module for the numpy-like particle physics friendly array, that supports nested, variable-sized data arrays.

Reviewer 2: ```suggestion
   - [awkward-array](https://github.com/scikit-hep/awkward-1.0) a Python module for a friendly manipulation of data structures common in Particle Physics, i.e., nested, variable-sized data, including arbitrary-length lists, records, mixed types, and missing data, using NumPy-like idioms.
```

   - [boost histogram](https://github.com/scikit-hep/boost-histogram) a python version of the C++14 library [Boost::Histogram](https://www.boost.org/doc/libs/1_77_0/libs/histogram/doc/html/index.html) for histogramming. 
   - [iminuit](https://github.com/scikit-hep/iminuit) a jupyter-friendly python interface for minimization package [Minuit2](https://root.cern.ch/doc/master/Minuit2Page.html).
   - [mplhep](https://github.com/scikit-hep/mplhep) a set of high energy physics helpers for [matplotlib](https://matplotlib.org/).
   - etc.
* [RooFit](https://root.cern/manual/roofit/) a fitting, fit model defining, plotting and "toy model" generation package supported by CERN ROOT team. Can be used in both C++ and python.

> * [zFit](https://github.com/zfit/zfit) a fitting and fit model defining python module integrated with [scikit-HEP project](https://github.com/scikit-hep) ecosystem.

Reviewer 2: ```suggestion
* [zfit](https://github.com/zfit/zfit) a fitting and fit model defining Python package well integrated withthe [Scikit-HEP project](https://github.com/scikit-hep) analysis ecosystem.
```

{% endcallout %}

> Discussions on the new analysis tools that might be useful for the LHCb community are held in the [work package 4](https://lhcb-dpa.web.cern.ch/lhcb-dpa/wp4/index.html) of the Data Processing and Analysis project (DPA). 

Reviewer 2: ```suggestion
Discussions on the new analysis tools that might be useful for the LHCb community are held in the [Work Package 4](https://lhcb-dpa.web.cern.ch/lhcb-dpa/wp4/index.html) of the Data Processing & Analysis project (DPA). 
```

### Analysis Preservation

> After one gets the necessary samples and writes the necessary macros and scripts to perform the analysis steps, like applying additional selections, fitting distributions, computing efficiencies and acceptances, etc. 

Reviewer: this sentence is not self-consistent, "after ..." what?

> After acquiring results, one has to have their work being review by physics working group (PWG) and then by collaboration. 

Reviewer: ```suggestion
After acquiring results, one has to have their work being reviewed by [physics working group](http://lhcb.web.cern.ch/lhcb_page/collaboration/organization/lhcb-conv/Physics_history_and_Sub-structure.html) (PWG) and then by the LHCb collaboration. 
```

A lesson on gitlab can be found [here](https://hsf-training.github.io/analysis-essentials/git/README.html).

However, it is not enough to just submit your code. 
It is important to write a set of instructions to execute the steps in the correct sequence.
You will quickly find yourself, that remembering how to run your own code even a couple of month after it was used for the last time can be really hard. 
Instructions can be written using [snakemake](https://snakemake.readthedocs.io/en/stable/) and supportive bash scripts.

Snakemake workflow management system allows you to write a pipeline for analysis that will automatically detect the hierarchy of the steps and changes in inputs (both samples and code). 
One should aim to start implementing the snakemake workflow from the beginning of analysis development.
Your snakemake workflow then should be committed to the gitlab repository with the final version of the code.

The samples you have produced should also be saved.
Analysis Productions help you to save a tested version of the DaVinci option files, that will help you to reproduce the ntuples with the same software version, that was specified in the original option files. 

> 

Reviewer: ```suggestion
```

The resulting samples have to be uploaded to your working group eos. 

> You should always preserve the package versions of the sofrware that you have being using.

Reviewer: ```suggestion
You should always preserve the package versions of the software that you have been using.
```

Reviewer: I am not sure this sends a clear message, but let the others comment. It would be good to say that _documenting_ your software versions in your analysis note is essential, but let's say I see little point in creating your own docker container just to store ROOT 6/24 which will anyway be accessible in other means.

Even if it is a commonly used tool like ROOT or numpy.

> This can be done by creating a [docker container](https://www.docker.com/), [miniconda enviroment](https://docs.conda.io/en/latest/miniconda.html) or using the [`lb-conda` environment](https://gitlab.cern.ch/lhcb-core/lbcondawrappers/-/blob/master/README.md).

Reviewer 3: It's better for people to use the conda-forge provided installers rather than the Anaconda provided ones if they don't already have conda insatalled:
```suggestion
This can be done by creating a [docker container](https://www.docker.com/), [conda enviroment](https://github.com/conda-forge/miniforge/#download) or using the [`lb-conda` environment](https://gitlab.cern.ch/lhcb-core/lbcondawrappers/-/blob/master/README.md).
```

# The analysis flow and analysis preservation

{% objectives "Learning Objectives" %}

* Learn about analysis tools
* Learn about analysis preservation 

{% endobjectives %}

> Most LHCb data analyses share similar structure and follow similar procedure pipeline.

Reviewer 2: ```suggestion
Most LHCb analyses share a similar structure and follow a similar procedure pipeline.
```

> As a user you start developing your analysis from retrieving the data and simulation samples that you need for your research.   

Reviewer 2: ```suggestion
As a user you often start developing your analysis by retrieving the data and simulation samples that you need.   
```

Giving you tools to get those samples is the main goal of the first analysis steps lessons.  

{% callout "Reminder from the data flow lesson" %}

> The output of the last stages of the [LHCb data flow](dataflow), namely Stripping (Run 1 and Run 2), Turbo stream (Run 2), Sprucing (Run 3) or Run 3 HLT2 lines can be accessed by user. 

Reviewer 2: ```suggestion
The output of the last stages of the [LHCb data flow](dataflow) - Stripping (Runs 1 and 2), Turbo stream (Run 2), Sprucing (Run 3) or Run 3 HLT2 lines - can be accessed by the user. 
```

Selections made at these last stages are also defined by users. 

> Their output is saved to disk and later is used to produce a data file that will contain only the information that interests users.

Reviewer 2: ```suggestion
Their outputs are saved to disk and later used to produce data files that will contain only the information that interests users.
```

This is done using the software package called DaVinci.

{% endcallout %}

### Getting data files

> After preselecting data either in the stripping, sprucing or triggering step, user can produce a ROOT file called _ntuple_, by running the DaVinci package.

Reviewer 2: ```suggestion
After preselecting data either in the Stripping, Sprucing or triggering step, users can produce ROOT files containing _ntuples_, running the DaVinci application.
```

> The ntuple is a data file that contains information about events or candidates in the data sample, like mass of a candidate or trigger decision flags.

Reviewer 2: ```suggestion
An ntuple is a (often complex) data structure typically stored within a (ROOT) file, which contains information about events or candidates in the data sample, such as the candidate mass or trigger decision flags.
```

> The ntuple is a data file that contains information about events or candidates in the data sample, like mass of a candidate or trigger decision flags.

Reviewer 2: I left just now a few suggestions but will have to pause. I will try and comment on the rest either later today or during the week. I see things that clearly need to be corrected - for example an ntuple is _not_ a file, it is rather a data structure typically stored within a (ROOT) file. Maybe worth linking to the ROOT manual for definitions of an ntuple? Else add a definition to our glossary https://github.com/lhcb/glossary and link to that.

Author: I opened a PR in glossary. As soon as merged I will link it here.

> For the data sample information is coming from reconstruction, but for the simulation sample you can request both the reconstructed values of observables/variables or generator-level values.

Reviewer 2: ```suggestion
For data samples the available information is solely that coming from the reconstruction, but for simulation samples you can request both the reconstructed values of observables/variables and the generator-level (truth) values.
```

The reconstructed values are the ones that passed the entire reconstruction pipeline with detector effects included and generator-level values are the original values that the event generator has produced.
Certain variables, for example, vertex quality or trigger decisions, can only be accessed on the reconstruction level.
This allows to study different reconstruction and detector effects for your analysis.

> DaVinci can be run locally at the lxplus (for small tasks) or at the CERN grid computing system via [ganga](davinci-grid) or via [Analysis Productions](analysis-productions). 

Reviewer 2: ```suggestion
DaVinci can be run locally on `lxplus` (for small tasks) or on the Grid computing system via [Ganga](davinci-grid) or via [Analysis Productions](analysis-productions). 
```

> Using ganga will give you more control over the jobs (little scrips and macros you run on the remote machines) and also will allow you to do way more than just running DaVinci. 

Reviewer 2: ```suggestion
Using Ganga will give you more control over the jobs (short scripts and macros you run on the remote machines) and also will allow you to do way more than just running DaVinci. 
```

Analysis Productions, however, are more user-friendly and provide a set of tests that ensure your jobs will be run as expected with minimal errors. 
Analysis Productions are also extremely important for the preservation of your analysis.
We will discuss the concept of analysis preservation a bit later in this lesson.

> Unless you want to run some really unusual ntuple jobs on ganga, you should always go for Analysis Productions. 

Reviewer 2: ```suggestion
Unless you want to run some really unusual ntuple jobs with Ganga, you should always go for Analysis Productions. 
```

> In first analysis steps we cover both running DaVinci on [ganga](https://lhcb.github.io/starterkit-lessons/first-analysis-steps/davinci-grid.html) and via [Analysis Productions](https://lhcb.github.io/starterkit-lessons/first-analysis-steps/analysis-productions.html).

Reviewer 2: ```suggestion
In first analysis steps we cover both running DaVinci on [Ganga](https://lhcb.github.io/starterkit-lessons/first-analysis-steps/davinci-grid.html) and via [Analysis Productions](https://lhcb.github.io/starterkit-lessons/first-analysis-steps/analysis-productions.html).
```

   - [boost histogram](https://github.com/scikit-hep/boost-histogram) a python version of the C++14 library [Boost::Histogram](https://www.boost.org/doc/libs/1_77_0/libs/histogram/doc/html/index.html) for histogramming. 
   - [iminuit](https://github.com/scikit-hep/iminuit) a jupyter-friendly python interface for minimization package [Minuit2](https://root.cern.ch/doc/master/Minuit2Page.html).
   - [mplhep](https://github.com/scikit-hep/mplhep) a set of high energy physics helpers for [matplotlib](https://matplotlib.org/).
   - etc.
* [RooFit](https://root.cern/manual/roofit/) a fitting, fit model defining, plotting and "toy model" generation package supported by CERN ROOT team. Can be used in both C++ and python.
* [zfit](https://github.com/zfit/zfit) a fitting and fit model defining Python package well integrated withthe [Scikit-HEP project](https://github.com/scikit-hep) analysis ecosystem.

{% endcallout %}

Discussions on the new analysis tools that might be useful for the LHCb community are held in the [Work Package 4](https://lhcb-dpa.web.cern.ch/lhcb-dpa/wp4/index.html) of the Data Processing & Analysis project (DPA). 

### Analysis Preservation

> When the samples are ready one can proceed with developing the necessary macros and scripts to perform the analysis steps, like applying additional selections, fitting distributions, computing efficiencies and acceptances, etc. 

Reviewer 2: ```suggestion
When the samples are ready one can proceed with developing the necessary macros and scripts to perform the analysis steps, such as applying additional selections, fitting distributions, computing efficiencies and acceptances, etc. 
```

Starting from the ntuples a typical analysis will consist of the following steps: 

1. Defining and applying selections, including kinematic, particle identification, multivariate analysis, signal/background separations by fitting.
2. Computing and applying callibrations and corrections. 

> Some of the calibrations, like [TrackCalib](https://gitlab.cern.ch/lhcb/Urania/tree/master/TrackCalib) and [PIDCalib2](https://gitlab.cern.ch/lhcb-rta/pidcalib2), are calibrated on the "standard candle" decays and can be used by the users directly. 

Reviewer 2: What do you mean? Rather calibration samples?

Author: Yes, I was thinking about how to explain it to a person with a very general knowledge of physics and I thought that the "standard candle" is sort of a universal term, but if you think it is even more confusing...

Reviewer 4: I think "standard candle" is indeed very confusing. I would use "calibration sample" and say that calibration samples are usually decays that can be reconstructed easily, are well understood and all together representative for most of the physics programme of LHCb. Maybe add that they are mainly used for tracking efficiency and PID corrections

Reviewer 5: TrackCalib and PIDCalib are examples of calibrations via  tag-and-probe approach, not a standard candle. A standard candle approach would be using a decay with a very well known branching fraction as efficiency normalisation. (think about the role of the J/psi -> mu mu / J/psi -> e e in R(K), and D -> K3pi / D -> Kpi)

Author: Thanks for clarification @laurentdufour  and comments @eduardo-rodrigues and @lukascalle. I have fixed it in the most generic way to not confuse the students more.

After acquiring the parameter(s) of interest, one has to have their work being reviewed by [physics working group](http://lhcb.web.cern.ch/lhcb_page/collaboration/organization/lhcb-conv/Physics_history_and_Sub-structure.html) (PWG) and then by the LHCb collaboration. 

One of the requirements for the successful review is that your analysis is fully preserved and can be reproduced by another LHCb physicist.  
Scientific integrity means having transparency on each step of the research as well as reproducibility of the results.
Both LHCb and CERN have guidelines which one should always follow when conducting research usign LHCb or CERN facilities.
LHCb guidelines are discussed and implemented in the [work package 6](https://lhcb-dpa.web.cern.ch/lhcb-dpa/wp6/index.html) of DPA.
CERN guideline can be found [here](https://analysispreservation.cern.ch/login?next=/).

The first step is to document well-used procedures, techniques and meticulously report the results of your analysis.
This is done by writing a clear and extensive Analysis Note, that is then available to the LHCb collaborators via the [CDS system](https://cds.cern.ch/collection/LHCb%20Analysis%20Notes).

> Secondly, your analysis code should be put at a [gitlab](https://gitlab.cern.ch/) repository in your physics working group's area. For example, the code of all recent analyses by the QEE working group can be found  [here](https://gitlab.cern.ch/LHCb-QEE).

Reviewer 2: ```suggestion
Secondly, your analysis code should be put in a [GitLab](https://gitlab.cern.ch/) repository in your physics working group's area. For example, the code of all recent analyses by the QEE working group can be found [here](https://gitlab.cern.ch/LHCb-QEE).
```

> Gitlab is useful for tracing your own changes and changes of your collaborators too, as well as, tracing possible bugs with gitlab-CI.

Reviewer 2: ```suggestion
GitLab is useful for tracing your own changes and changes of your collaborators too, as well as, tracing possible bugs with the GitLab CI.
```

> A lesson on gitlab can be found [here](https://hsf-training.github.io/analysis-essentials/git/README.html).

Reviewer 2: ```suggestion
Refer to [this lesson on GitLab](https://hsf-training.github.io/analysis-essentials/git/README.html) for more.
```

You will quickly find yourself, that remembering how to run your own code even a couple of months after it was used for the last time can be really hard. 
Instructions can be written using [snakemake](https://snakemake.readthedocs.io/en/stable/) and supportive bash scripts.

Snakemake workflow management system allows you to write a pipeline for analysis that will automatically detect the hierarchy of the steps and changes in inputs (both samples and code). 
One should aim to start implementing the snakemake workflow from the beginning of analysis development.
Your snakemake workflow should then be committed to the gitlab repository with the final version of your code.

The samples you have produced should also be saved.
Analysis Productions help you to save a tested version of the DaVinci option files, that will help you to reproduce the ntuples with the same software version, that was specified in the original option files. 

The ntuples used for your analysis have to be uploaded to the [eos area](eos-storage) of your working group. 

You should always preserve the package versions of the sofrware that you have being using.

> Even if it is a commonly used tool like ROOT or numpy.

Reviewer 2: ```suggestion
Even if it is a commonly used tool such as ROOT or NumPy.
```

This can be done by creating a [docker container](https://www.docker.com/), [conda enviroment](https://github.com/conda-forge/miniforge/) or using the [`lb-conda` environment](https://gitlab.cern.ch/lhcb-core/lbcondawrappers/-/blob/master/README.md).

> The latter uses the conda environments installed on [Cern virtual machine file system](https://cernvm.cern.ch/fs/) or CVMFS and is supported by LHCb community. 

Reviewer 2: ```suggestion
The latter uses the conda environments installed on the [CERN virtual machine file system](https://cernvm.cern.ch/fs/), aka CVMFS, and is supported by the LHCb community. 
```

# The analysis flow and analysis preservation

{% objectives "Learning Objectives" %}

* Learn about analysis tools
* Learn about analysis preservation 

{% endobjectives %}

Most LHCb analyses share a similar structure and follow a similar procedure pipeline.
As a user you often start developing your analysis by retrieving the data and simulation samples that you need.   
Giving you tools to get those samples is the main goal of the first analysis steps lessons.  

{% callout "Reminder from the data flow lesson" %}

The output of the last stages of the [LHCb data flow](dataflow) - Stripping (Runs 1 and 2), Turbo stream (Run 2), Sprucing (Run 3) or Run 3 HLT2 lines - can be accessed by the user. 
Selections made at these last stages are also defined by users. 
Their outputs are saved to disk and later used to produce data files that will contain only the information that interests users.
This is done using the software package called DaVinci.

{% endcallout %}

### Getting data files

> <<<<<<< HEAD

Author: Thanks! fixed
